language/ensure_spec.rb

```diff
@@ -124,3 +124,56 @@
     @obj.explicit_return_in_method_with_ensure.should == :ensure
   end
 end
+
+describe "An ensure block inside a class" do
+  before :each do
+    ScratchPad.record []
+  end
+
+  it "is allowed" do
+    lambda {
+      eval <<-ruby
+        class EnsureInClassExample
+          raise
+        ensure
+          ScratchPad << :ensure
+        end
+      ruby
+    }.should raise_error(RuntimeError)
+
+    ScratchPad.recorded.should == [:ensure]
+  end
+end
```

Comment:
@eregon 
Here I added only one happy path. There are another two similar sections `"An ensure block inside a begin block"` and `"An ensure block inside a method"`. Both of them contain the same five test cases. So I wonder if I should test new section `"An ensure block inside a class"` the same way.

Maybe I can generalize the question. What is the intention of the whole project (I know - `practical tool`, `test the behavior`, `It helps Ruby implementations` but still) and how comprehensive should tests be?

I may assume following levels:
- just smoky tests, we test only happy path
- we test behaviour, we declare the behaviour, it's a runnable specification
- we test everything, every edge case, every combination of parameters/constructions/etc

There are all kinds of specs listed above in the project. So I am curious what is the goal now? How should we test now?

Reply:
@andrykonchin 
The goal is essentially your level 2: `we test behaviour, we declare the behaviour, it's a runnable specification`.

Level 3 `we test everything, every edge case, every combination of parameters/constructions/etc` seems unrealistic, because there is an infinite number of combinations.
However, we should strive to spec a representative subset of parameters, including edge cases.

Level 1 `just smoky tests, we test only happy path` is really not ideal, as it leaves parts of the feature untested.
However, it is always better to have some specs than none at all, so I generally accept PRs adding specs even if they are not "complete" or covering every edge case.

Reply:
For this specific case, I think it would be better to test `ensure` with and without exception, making sure it executes and returns the expected value/rethrows the exception. I think it's OK to have 2 examples instead of 4 here, by testing both that the `ensure` clause is executed and that it returns the expected value/rethrows the exception in the same example.
As an aside, when throwing an exception, it's better to use a specific exception class, so that other exceptions thrown by the Ruby implementation cannot be confused with it (happens quite frequently in my experience).

@andrykonchin  Could you try to improve this PR with this guideline?

Reply:
@eregon done